src/payment/strategies/adyenv2/adyenv2-payment-strategy.ts

```diff
@@ -142,83 +153,23 @@ export default class AdyenV2PaymentStrategy implements PaymentStrategy {
-                            this._store.dispatch(this._paymentActionCreator.submitPayment({
-                                ...payment,
-                                paymentData: {
-                                    ...payment.paymentData,
-                                    shouldSaveInstrument,
-                                },
-                            })))
-                        .catch(error => {
-                            if (!(error instanceof RequestError) || !some(error.body.errors, { code: 'three_d_secure_required' })) {
-                                return Promise.reject(error);
-                            }
-
-                            if (error.body.three_ds_result.code === ResultCode.ChallengeShopper) {
-                                return this._handle3DS2Challenge(error.body.three_ds_result, payment.methodId)
-                                    .then((payment: Payment) =>
-                                        this._store.dispatch(this._paymentActionCreator.submitPayment({
-                                            ...payment,
-                                            paymentData: {
-                                                ...payment.paymentData,
-                                                shouldSaveInstrument,
-                                            },
-                                        }))
-                                    );
-                            }
-
-                            return Promise.reject(error);
-                        });
-                }
-
-                if (error.body.three_ds_result.code === ResultCode.ChallengeShopper) {
-                    return this._handle3DS2Challenge(error.body.three_ds_result, payment.methodId)
-                        .then((payment: Payment) =>
-                            this._store.dispatch(this._paymentActionCreator.submitPayment({
-                                ...payment,
-                                paymentData: {
-                                    ...payment.paymentData,
-                                    shouldSaveInstrument,
-                                },
-                            }))
-                        );
-                }
-
-                if (some(error.body.errors, { code: 'additional_action_required' })) {
-                    return this._handleFromAction(error.body.provider_data);
-                }
-
-                return new Promise(() => {
-                    this._formPoster.postForm(error.body.three_ds_result.acs_url, {
-                        PaReq: error.body.three_ds_result.payer_auth_request,
-                        TermUrl: error.body.three_ds_result.callback_url,
-                        MD: error.body.three_ds_result.merchant_data,
-                    });
-                });
-            });
+            .catch(error => this._processPayment(error, shouldSaveInstrument))
+            .catch(error => this._processPayment(error, shouldSaveInstrument));
```

Comment:
is not very obvious that if something fails (catching the error) then you proceed to process the payment. Also is not obvious why you are calling it twice. Might be worth trying to make the code a bit more self explanatory

```js
.catch(error => {
  if (error.type == 'additional_action_required') {
    this._processAdditionalActionRequired(error, shouldSaveInstrument));
  }
  throw error;
})`
```

If the idea behind is to call itself as long as there is a `additional_action_required`, maybe considering something recursive instead of hard coding it twice?